### ISSUE TITLE
Table name adaptions to v0.2 of oedb

### ISSUE THREAD

**Issue by gplssm** [openego/ego.io#8]
We need to adapt new table names to v0.2. We do this in `refactor/adaptions-of-tables-names`.

Do we want to change names of ORM-classes or do we just adapt underlying table names

**Comment by ulfmueller** [comment]
I would say we should rename both. But doesnt @MGlauer had a script to automatically create the ego.io definitions from the oedb structure?

**Comment by gplssm** [comment]
Good point!

I suggest, we wait for a reply of @MGlauer. Meanwhile I adapt names of table required in the _data processing_.

**Comment by ulfmueller** [comment]
@MGlauer Can we update the ego.io definitions in an automatized way like it was done in the first place?
(see above)

**Comment by nesnoj** [comment]
@MGlauer Any updates on this? I'd like to import new table defs of model draft but there's none..

Update by @MGlauer (via Mail): He aims to finish it by the end of this week..

**Comment by MGlauer** [comment]
A new schema definition is available in  e96c23c.

I guess there are still some manual adaptations required. But this is the general structure.

**Comment by nesnoj** [comment]
Cheers!

I got some issues on the new defs:

1. In model_draft ARRAY is tried to import from `sqlalchemy.dialects.postgresql.base`. The right place is `sqlalchemy.dialects.postgresql`
2. Some psql-specific types are missing: ARRAY, DOUBLE_PRECISION, INTEGER, NUMERIC, TEXT
3. In EgoScenarioLog the [attribute metadata](https://github.com/openego/ego.io/blob/e96c23cc36ff5cf527ea6c46e2083a01fc57a3a1/egoio/db_tables/model_draft.py#L1869) is used - this throws:
```
"Attribute name 'metadata' is reserved "
sqlalchemy.exc.InvalidRequestError: Attribute name 'metadata' is reserved for the MetaData instance when using a declarative base class.
```
The attribute collides with the MetaData instance [here](https://github.com/openego/ego.io/blob/e96c23cc36ff5cf527ea6c46e2083a01fc57a3a1/egoio/db_tables/model_draft.py#L11). So to avoid renaming the table column "metadata" in `model_draft.ego_scenario_log` (and accessing scripts) I mapped the table name using "meta_data". 

I fixed 1-3 manually in c49b46c3dfb4446e6446f4233877728b6907072f.

How can we deal with 3. when using your script? (without renaming the column)

**Comment by nesnoj** [comment]
@MGlauer: Could you provide your autogen script? I'd like to add some defs...and additionally you might get support on bugs :)

**Comment by MGlauer** [comment]
There is no real magic to this script. I am just using [sqlacodegen](https://pypi.python.org/pypi/sqlacodegen) in a simple bash-script. I can add it to the dictionary, if you wish, but I guess you just want to call `sqlacodegen`  yourself
```
#!/bin/bash

db_user=""
db_password=""
db_host=""
db_port="5432"
db="oedb"
path="egoio/db_tables/"

if [ -d "egoio/db_tables" ]; then
    schemas=("demand" "economic" "emission" "environmental" "grid" "openstreetmap" "political_boundary" "reference" "scenario" "social" "supply" "weather" "model_draft")
    for schema in ${schemas[*]}
    do 
        {
            echo "Write table schemas for $schema"
            sqlacodegen postgres://$db_user:$db_password@$db_host:$db_port/$db --schema=$schema --outfile=$path$schema.py 
        } || { 
            echo "Error in schema ${schemas[$i]}"
            mv $path$schema.py  {$path$schema}_err.py 
        }
    done
else
    echo "Folder egoio/db_tables does not exist"
fi
```

**Comment by nesnoj** [comment]
Thanks for sharing!
I gave it try in a temp branch but unfortunately it doesn't produce defs, only files with content
```
from sqlalchemy import MetaData
metadata = MetaData()
```
No errors occurred during execution and permissions do not seem to be the reason (same output on a temp schema created by me). Do you know this issue?

If you cant spend time on this right now, as a temporary solution it would be great if you just run it on a temp branch since I only need some of the defs - I'll delete it afterwards.

**Comment by MGlauer** [comment]
Oh, sorry, seems like I accidently added a `}` when copying it. I edited the source on top.

**Comment by nesnoj** [comment]
It's strange that the script didn't throw an exception..
Great, it's working now, thank you.

**Comment by nesnoj** [comment]
**Note for other users of the above script:**
Geom columns are translated into NullTypes since sqlacodegen (v1.1.6) doesn't support geometry columns by default. To fix this you have to add the import of Geoalchemy2's Geometry type to codegen.py (e.g. in /usr/local/lib/python3.4/dist-packages/sqlacodegen/codegen.py): `from geoalchemy2 import Geometry` ([source](http://stackoverflow.com/questions/34594165/sqlacodegen-issue-with-geometry-type)).

Furthermore, there're some warnings concerning constraints (don't care):
```
...
Write table schemas for economic
/usr/local/lib/python3.4/dist-packages/sqlalchemy/dialects/postgresql/base.py:2783: SAWarning: Skipped unsupported reflection of expression-based index enforce_spatially_unique_ioer_urban_share_industrial_rast
  % idx_name)
/usr/local/lib/python3.4/dist-packages/sqlalchemy/dialects/postgresql/base.py:2783: SAWarning: Skipped unsupported reflection of expression-based index ioer_urban_share_industrial_gix
  % idx_name)
Write table schemas for emission
...

```

**Comment by ulfmueller** [comment]
@nesnoj did you manage get the adaptation script running? @s3pp has used it successfully. Maybe he can help.

**Comment by nesnoj** [comment]
Yupp, worked out - further discussions see #14

**Comment by ulfmueller** [comment]
can we have an update run in order to include new class definitions of db tables? @nesnoj 
I havent used sqlacodegen. Can I easily use it, like stated above?

**Comment by nesnoj** [comment]
There're some drawbacks, one is stated above. I'm think I did not document them all..
If you can wait until tomorrow morning I could take care of it and furthermore give a more detailed list of the issues to take care of...otherwise, feel free :) (but please add new branch)

**Comment by ulfmueller** [comment]
Yes, I can wait until today ;) It would be very nice if you could take care of it and create a little documentation in order to reproduce that work easily! Thanks a lot for offering! If you want, you do only the second part and someone else (like me) tries it out...
Of course within a new branch!

**Comment by nesnoj** [comment]
I built a new release of ego.io for dp 0.2 release in branch [schema_structure_dp_v0.2](https://github.com/openego/ego.io/tree/refactor/schema_structure_dp_v0.2).

This is a how-to which we could move to the wiki as a help for future releases.

**How to build new definitions:**

0. **Create new branch**
1. **Add Geoalchemy2's Geometry types to sqlacodegen**
Geom columns are translated into NullTypes since sqlacodegen (v1.1.6) doesn't support geometry columns by default. To fix this you have to add the import of Geoalchemy2's Geometry type to codegen.py locally on your installation (e.g. in /usr/local/lib/python3.4/dist-packages/sqlacodegen/codegen.py): `from geoalchemy2 import Geometry` ([source](http://stackoverflow.com/questions/34594165/sqlacodegen-issue-with-geometry-type)).

2. **Run [codegen](https://github.com/openego/ego.io/blob/5a6038cd2ce76a94ac60963b2697fd3cdaeac7c4/egoio/tools/sqlacodegen_oedb.sh)**
Add user, pass, server and port to script and run.
I got some warnings which I ignored (not sure if there will result any problems..):
```
Write table schemas for demand
Write table schemas for economic
Write table schemas for emission
Write table schemas for environmental
/usr/local/lib/python3.4/dist-packages/sqlalchemy/dialects/postgresql/base.py:2783: SAWarning: Skipped unsupported reflection of expression-based index ecjrc_glc2000_rast_idx
  % idx_name)
/usr/local/lib/python3.4/dist-packages/sqlalchemy/dialects/postgresql/base.py:2783: SAWarning: Skipped unsupported reflection of expression-based index noaa_globe_rast_idx
  % idx_name)
Write table schemas for grid
Write table schemas for openstreetmap
/usr/local/lib/python3.4/dist-packages/sqlalchemy/dialects/postgresql/base.py:2790: SAWarning: Predicate of partial index osm_deu_ways_idx ignored during reflection
  % idx_name)
/usr/local/lib/python3.4/dist-packages/sqlalchemy/dialects/postgresql/base.py:2790: SAWarning: Predicate of partial index osm_deu_rels_idx ignored during reflection
  % idx_name)
Write table schemas for political_boundary
Write table schemas for reference
Write table schemas for scenario
Write table schemas for social
Write table schemas for supply
Write table schemas for weather
Write table schemas for model_draft
/usr/local/lib/python3.4/dist-packages/sqlalchemy/dialects/postgresql/base.py:2783: SAWarning: Skipped unsupported reflection of expression-based index enforce_spatially_unique_ioer_urban_share_industrial_rast
  % idx_name)
/usr/local/lib/python3.4/dist-packages/sqlalchemy/dialects/postgresql/base.py:2783: SAWarning: Skipped unsupported reflection of expression-based index ioer_urban_share_industrial_gix
  % idx_name)
```
**Do not forget no stash your credentials!**

3. **Bugfix: Add psql-specific modules / data types**
Some datatypes are not written by sqlacodegen which are to be added manually (139780a7a976f8b6164fcf9d2a68ca71b1e54249).

4. **Bugfix: Map reserved 'metadata' to 'meta_data**
In EgoScenarioLog the attribute [`metadata`]() is used. The attribute collides with the MetaData instance [here](). So to avoid renaming the table column "metadata" in `model_draft.ego_scenario_log` (and accessing scripts) I mapped the table name using "meta_data" (bcf51a74c9fff659dc865db7f2741a910c976844).

5. **Bugfix: Tables ending with `Bus` are translated into `Bu`**
No idea where this *Bu bug (see [here](https://github.com/openego/ego.io/issues/14#issuecomment-291530214)) comes from -> add missing `s`
Occurrences: `model_draft.py` and `grid.py`
(573e7563231bd1c32ca45cd29d12a338e8e26a17, 5fcc006bf8fc8dc467466faebd4803e161574dfd, 7c175cbf4f994d72fccfafdce2b9f8645121d518)

6. **Eventually adjust [setup.py](https://github.com/openego/ego.io/blob/5a6038cd2ce76a94ac60963b2697fd3cdaeac7c4/setup.py)** (5a6038cd2ce76a94ac60963b2697fd3cdaeac7c4)
7. **Test it**
8. **Create pull request**
It works well with DINGO and doesn't throw any exceptions here, but if you stumble upon any issues, please document them here so we can add required modifications to the list above. If everything is fine, please merge the corresponding [pull request](https://github.com/openego/ego.io/pull/20). DONE
9. **Release via PyPI**
If there're no objections and the pull request has been merged, I will release it via PyPI.

**Comment by nesnoj** [comment]
Merge was done (#20), please use `dev` as I will delete the branch `refactor/schema_structure_dp_v0.2` within the next days.

**Comment by nesnoj** [comment]
I created a [repo wiki](https://github.com/openego/ego.io/wiki) and added the How-To.
2 more Questions:
1. We need a How-to-release (example from [oemof](https://github.com/oemof/oemof/wiki/How-to-release))
2. How long are we gonna wait from pull request to merge? (see [wikipage](https://github.com/openego/ego.io/wiki/How-to-update-database-definitions) no. 8)

I got the feeling that there's some more stuff to harmonize - It seems we need some more guidelines on branching, releasing, ... :book: 
A discussion point for next WebCo?

**Comment by eosram** [comment]
I just now stumbled on an error. Three columns --> attributes did not find their way into the class definition. I tested running the script again, still missing. Fixed in in 6fc170aacf3d0acea896c0d7132ab4facdf4e9e8. I also updated the wiki [accordingly](https://github.com/openego/ego.io/wiki/How-to-update-database-definitions/_compare/97f221c6936c08b75eb1ee5f9ea2a48ca0cef352...fff9b606116791f14ed399a3cd9272967b0ba00e).
